src/lib.rs

```diff
@@ -1,11 +1,279 @@
+        }
+
+        impl<W> $name<W> {
+            fn new<T: ByteOrder>(w: W, value: $ty) -> Self {
+                let mut writer = $name {
+                    buf: [0; $bytes],
+                    written: 0,
+                    dst: w,
+                };
+                T::$writer(&mut writer.buf[..], value);
+                writer
+            }
+        }
+
+        impl<W> Future for $name<W>
+        where
+            W: io::AsyncWrite,
+        {
+            type Output = io::Result<()>;
+            fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+                if self.written == $bytes as u8 {
+                    return Poll::Ready(Ok(()));
+                }
+
+                // we need this so that we can mutably borrow multiple fields
+                // it is safe as long as we never take &mut to dst (since it has been pinned)
+                // unless it is to place it in a Pin itself like below.
+                let mut this = unsafe { self.get_unchecked_mut() };
+                let mut dst = unsafe { Pin::new_unchecked(&mut this.dst) };
+
+                while this.written < $bytes as u8 {
+                    this.written += match dst
+                        .as_mut()
+                        .poll_write(cx, &this.buf[this.written as usize..])
+                    {
+                        Poll::Pending => return Poll::Pending,
+                        Poll::Ready(Err(e)) => return Poll::Ready(Err(e.into())),
+                        Poll::Ready(Ok(n)) => n as u8,
+                    };
+                }
+                Poll::Ready(Ok(()))
+            }
+        }
+    };
+}
+
+macro_rules! writer8 {
+    ($name:ident, $ty:ty) => {
+        #[doc(hidden)]
+        pub struct $name<W>(W, $ty);
+        impl<W> Future for $name<W>
+        where
+            W: io::AsyncWrite,
+        {
+            type Output = io::Result<()>;
+            fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+                let this = unsafe { self.get_unchecked_mut() };
+                let dst = unsafe { Pin::new_unchecked(&mut this.0) };
+                let buf = [this.1 as u8];
+                match dst.poll_write(cx, &buf[..]) {
+                    Poll::Pending => Poll::Pending,
+                    Poll::Ready(Err(e)) => Poll::Ready(Err(e.into())),
+                    Poll::Ready(Ok(0)) => Poll::Pending,
+                    Poll::Ready(Ok(1)) => Poll::Ready(Ok(())),
+                    Poll::Ready(Ok(_)) => unreachable!(),
+                }
+            }
+        }
+    };
+}
+
+macro_rules! write_impl {
+    (
+        $(#[$outer:meta])*
+        fn $name:ident(&mut self, n: $ty:ty) -> $($fut:tt)*
+    ) => {
+        $(#[$outer])*
+        #[inline]
+        fn $name<'a, T: ByteOrder>(&'a mut self, n: $ty) -> $($fut)*<&'a mut Self> where Self: Unpin {
+            $($fut)*::new::<T>(self, n)
+        }
+    }
+}
+
+/// Tokio traits feature gate
```

Comment:
This will end up being the documentation for the `tokio` mod, which is probably not what you wanted.